src/cargo/ops/fix.rs

```diff
@@ -578,70 +586,92 @@ fn log_failed_fix(stderr: &[u8]) -> Result<(), Error> {
+    /// `FixArgs`.
     other: Vec<OsString>,
-    rustc: Option<PathBuf>,
+    /// Path to the `rustc` executable.
+    rustc: PathBuf,
+    /// Console output flags (`--error-format`, `--json`, etc.).
+    ///
+    /// The normal fix procedure always uses `--json`, so it overrides what
+    /// Cargo normally passes when applying fixes. When displaying warnings or
+    /// errors, it will use these flags.
     format_args: Vec<String>,
 }
 
-enum PrepareFor {
-    Next,
-    Edition(Edition),
-    None,
-}
-
-impl Default for PrepareFor {
-    fn default() -> PrepareFor {
-        PrepareFor::None
-    }
-}
-
 impl FixArgs {
     fn get() -> Result<FixArgs, Error> {
-        let mut ret = FixArgs::default();
-
-        ret.rustc = env::args_os().nth(1).map(PathBuf::from);
+        let rustc = env::args_os()
+            .nth(1)
+            .map(PathBuf::from)
+            .ok_or_else(|| anyhow::anyhow!("expected rustc as first argument"))?;
+        let mut file = None;
+        let mut enabled_edition = None;
+        let mut other = Vec::new();
+        let mut format_args = Vec::new();
 
         for arg in env::args_os().skip(2) {
             let path = PathBuf::from(arg);
             if path.extension().and_then(|s| s.to_str()) == Some("rs") && path.exists() {
-                ret.file = Some(path);
+                file = Some(path);
                 continue;
             }
             if let Some(s) = path.to_str() {
                 if let Some(edition) = s.strip_prefix("--edition=") {
-                    ret.enabled_edition = Some(edition.parse()?);
+                    enabled_edition = Some(edition.parse()?);
                     continue;
                 }
                 if s.starts_with("--error-format=") || s.starts_with("--json=") {
                     // Cargo may add error-format in some cases, but `cargo
                     // fix` wants to add its own.
-                    ret.format_args.push(s.to_string());
+                    format_args.push(s.to_string());
                     continue;
                 }
             }
-            ret.other.push(path.into());
+            other.push(path.into());
         }
-        if let Ok(s) = env::var(PREPARE_FOR_ENV) {
-            ret.prepare_for_edition = PrepareFor::Edition(s.parse()?);
+        let file = file.ok_or_else(|| anyhow::anyhow!("could not find .rs file in rustc args"))?;
+        let idioms = env::var(IDIOMS_ENV).is_ok();
+
+        let prepare_for_edition = if let Ok(s) = env::var(PREPARE_FOR_ENV) {
+            Some(s.parse()?)
         } else if env::var(EDITION_ENV).is_ok() {
-            ret.prepare_for_edition = PrepareFor::Next;
-        }
+            match enabled_edition {
+                None | Some(Edition::Edition2015) => Some(Edition::Edition2018),
+                Some(Edition::Edition2018) => Some(Edition::Edition2021),
+                Some(Edition::Edition2021) => Some(Edition::Edition2021),
+            }
```

Comment:
While technically somewhat unrelated to this PR, could this `match` be encapsulated in a method on `Edition` for easy editing to this in the future?